### PR TITLE
Cutouts

### PR DESCRIPTION
**Pull request by ColeFMeldorf**
In this PR I do a few things.
The main one is add an ability to create a cutout from an OpenUniverse2024FitsImage.
I also added the ability to get the image shape, the coordinate center of the image (useful for SMP) and the band.

### CODE REVIEW

File: snappl/image.py [hunk 1]
```diff
@@ -252,3 +298,80 @@ def get_wcs( self ):
...
+            with fits.open(self.inputs.path) as hdul:
+                return hdul[1].header
+        else:
+            return self._header
+
+    @property
+    def image_shape(self):
+        """Get the shape of the image."""
+        if not self._is_cutout:
+            return (4088, 4088)
```
Comment (wmwv) [inline snappl/image.py:314]: Don't do this (as you said in your comment).  Ask the image its size.  You should have the nx, ny from the metadata, or at worst you can ask the data array its size (although make sure you agree on the dimension).

Comment (wmwv) [reply snappl/image.py:314]: Specifically, it should just be
self._header["NX"]
self._header["NY"]

**Review by wmwv — CHANGES_REQUESTED** [review]
Write a test to exercise this function.
1. Cover the intended use case
2. Cover corner cases (being right on the edge)
3. Cover and define behavior of asking for a cutout that extends outside the image.

File: snappl/image.py (hunk 1 above)
Comment (ColeFMeldorf) [reply snappl/image.py:314]: Re: your test ideas 2 and 3, astropy Cutout2D with the 'strict' option automatically fails if the cutout is outside of the image at all. Should my test just check that it fails?

Comment (rknop) [reply snappl/image.py:314]: That's fine.  Probably document in the base Image class that cutout is not guaranteed to work if the cutout isn't entirely contained by the image.

**Review by rknop — COMMENTED** [review]
One other very minor change you're going to need to make is create something for the change log.

To do this, you need `towncrier` installed, which can be simply installed with `pip`.  Then, in the top level of your checkout, run
```
towncrier create 5.snappl
```

Type a short one-line description of what this PR does.

Then do 
```
git add changes/5.snappl.rst
```

followed by the usual commit and push.  If you do this right, that will cause the currently-failing test to pass.

(All of this will be documented in the package template once we get a few more PRs for *that* package in.)

**Comment by ColeFMeldorf** [comment]
All changes addressed! Let me know what you both think! @rknop @wmwv

File: snappl/tests/test_opensim2024fitsimage.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+sys.path.insert(0, str(pathlib.Path(__file__).parent/"extern/snappl"))
+# End of lines that will go away once we do this right. (<-- From Rob.)
+
+import numpy as np
+from snappl.image import OpenUniverse2024FITSImage
+import astropy
+
+
+def test_get_cutout():
+    roman_path = '/hpc/group/cosmology/OpenUniverse2024'
```
Comment (wmwv) [inline snappl/tests/test_opensim2024fitsimage.py:17]: This will fail on any other system.  Consider using ROMAN_SNPIT or consider adding a test file directly into the snappl repo.

File: snappl/tests/test_opensim2024fitsimage.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    cutout = image.get_cutout(ra, dec, 5)
+    comparison_cutout = np.load(str(pathlib.Path(__file__).parent/'image_test_data/test_cutout.npy'),
+                                allow_pickle=True)
+    message = "The cutout does not match the comparison cutout"
+    assert np.array_equal(cutout._data, comparison_cutout), message
+    # I am directly comparing for equality here because these numbers should
+    # never actually change, provided the underlying image is unaltered. -Cole
+
+    # Now we intentionally try to get a no overlap error.
+    try:
```
Comment (wmwv) [inline snappl/tests/test_opensim2024fitsimage.py:36]: Use `pytest.raises`

https://docs.pytest.org/en/7.1.x/how-to/assert.html#assertions-about-expected-exceptions

**Review by wmwv — CHANGES_REQUESTED** [review]
1. Use pytest.raises
2. Plan with @rknop for having test data.

File: snappl/image.py [hunk 4]
```diff
@@ -311,12 +311,13 @@ def _get_header(self):
     def image_shape(self):
         """Get the shape of the image."""
         if not self._is_cutout:
-            return (4088, 4088)
-            # This is a nasty hardcode but these images will always be
-            # this size.....? -Cole
+            self._header = self.get_header()
```
Comment (wmwv) [inline snappl/image.py:314]: 1. Why do these paths have to be different?  What's wrong with self.data.shape?  Where does that get set?

2. Why isn't `self._header` already defined?

3. Calling this function defines ._header and ._image_shape as a side effect.  Is that intentional.

**Review by wmwv — CHANGES_REQUESTED** [review]
Check with @rknop about the philosophy of whether snappl is supposed to give you an object that just keeps tracks of where the file is, or if it loads header, or loads the data array.  Are these different states of the same object or different objects?

File: snappl/tests/test_opensim2024fitsimage.py (hunk 2 above)
Comment (rknop) [reply snappl/tests/test_opensim2024fitsimage.py:17]: Yeah, if you look, you'll see that I have a subdirectory `config_test_data` where I stuck some `.yaml` files.  It's OK to put a FITS file or two into the repo -- though if you crop it first so it's small to start with, that's better.  You can make a generic `test_data` subdirectory if you want.

File: snappl/image.py (hunk 4 above)
Comment (rknop) [reply snappl/image.py:314]: To answer the question in the parent comment : different states of the same object.  Sort of.  Except that "the header" may not be well defined, given that we need to support various different image formats.  So, the parent class isn't going to necessarily know anything about headers, it will just have functions for getting stuff that, at least from FITS, you get from a FITS header.

File: snappl/image.py [hunk 5]
```diff
@@ -252,3 +298,81 @@ def get_wcs( self ):
             with fits.open( self.inputs.path ) as hdul:
                 self._wcs = AstropyWCS( hdul[1].header )
         return self._wcs
+
+    def _get_header(self):
+        """Get the header of the image."""
+        if self._header is None:
+            with fits.open(self.inputs.path) as hdul:
+                return hdul[1].header
```
Comment (rknop) [inline snappl/image.py:306]: Replace this with `self._header = hdul[1].header` so that the header is cached and we never have to read it again.

File: snappl/image.py [hunk 6]
```diff
@@ -252,3 +298,81 @@ def get_wcs( self ):
...
+        else:
```
Comment (rknop) [inline snappl/image.py:307]: Then remove this else.  So, the function starts with `if self._header is None:`, loads the header if that passes.  At the end, returns `self._header` no matter what.

File: snappl/image.py (hunk 4 above)
Comment (rknop) [reply snappl/image.py:314]: Re: the deal with `_header`.  These image classes in general do "lazy loading" of image data.  That is, they don't actually read it from disk until you need it.  Same thing with the header -- though header is FITS specific.

Because there is a cost to opening and reading files, and especially becaused FITS headers aren't large, we just cache it the first time we read it.

Cole, make the changes I noted above in `_get_header`.  Then, here, you don't  need to do anything with the return value, just call `self._get_header()`.

File: snappl/image.py [hunk 7]
```diff
@@ -146,6 +150,43 @@ def get_wcs( self ):
         raise NotImplementedError( f"{self.__class__.__name__} needs to implement get_wcs" )
 
 
+    def get_cutout(self, ra, dec):
```
Comment (rknop) [inline snappl/image.py:153]: Two things.  First, the signature of this function should be the same as what you have below in `OpenUniverse2024FITSImage`, meaning you need a `size` argument.

Thinking about it, though, the first function for getting a cutout that I'd think about would pass pixel coordinates, not ra / dec.  I might rename this to `get_cutout_at_ra_dec` or some such, and have `get_cutout` take `(x, y, xsize, ysize)`, or maybe `(x0, x1, y0, y1)`. 

You need to document all of the parameters.  In particular, just looking the definition of `OpenUniverse2024FITSImage.get_cutout` below, I don't know what the units of size are supposed to be.

File: snappl/image.py (hunk 4 above)
Comment (rknop) [reply snappl/image.py:314]: Finally, the advantage of having a `image_shape` function that's not just `data.shape` is that there may be a circumstance where you need the shape of the image but don't need the image data, so there's no need to do the file i/o and use the memory to store the data array just to get the shape.

File: snappl/tests/test_opensim2024fitsimage.py (hunk 2 above)
Comment (ColeFMeldorf) [reply snappl/tests/test_opensim2024fitsimage.py:17]: Done!

File: snappl/image.py (hunk 7 above)
Comment (ColeFMeldorf) [reply snappl/image.py:153]: Sorry I don't fully understand this:

> Cole, make the changes I noted above in _get_header. Then, here, you don't need to do anything with the return value, just call self._get_header().

Are you saying other than the changes above nothing else needs to be changed here? Or what do you mean "don't do anything with the return value'? @rknop

**Comment by ColeFMeldorf** [comment]
Other than one small question, all changes addressed. Let me know what you think @rknop @wmwv .
